### PR TITLE
Site Editor: Fix gray screen on refresh when editing pages and posts

### LINKED ISSUE #28315: Site Editor: Refreshing when editing a post or page is causing a gray screen
## Step-by-step reproduction instructions
1. Open site editor
2. Open navigation panel
3. Navigate to "Pages"
4. Open "Sample page" (or any page you have)
5. Press refresh
6. Gray screen 😢  (parameters are removed from the URL after loading on refresh)

## Expected behaviour
Page to load

## Actual behaviour
Nothing is loaded

## Code
https://github.com/WordPress/gutenberg/blob/97f0a489276c85af2bf3c764bc1fb315b5ecb0bd/packages/edit-site/src/components/url-query-controller/index.js

## Screenshots or screen recording (optional)
https://user-images.githubusercontent.com/2256104/105016576-1796ce80-5a43-11eb-8034-7d698522b579.mov

### PR DESCRIPTION
**Pull request by david-szabo97**
Fixes #28315

## Description
`setPage` action issued wrong requests when passing `postType` and `postId` only (no path specified). Let's use the right URL based on the `postType`.

Note: `path` should be absolute here to avoid issuing requests to `admin.php`.
## How has this been tested?

1. Open site editor
2. Open navigation panel
3. Navigate to "Pages"
4. Open "Sample page" (or any page you have)
5. Press refresh
6. Make sure it loads the same page
7. Repeat with a post too

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/edit-site/src/store/actions.js [hunk 1]
```diff
@@ -135,7 +135,11 @@ export function setHomeTemplateId( homeTemplateId ) {
  */
 export function* setPage( page ) {
 	if ( ! page.path && page.context?.postId ) {
-		page.path = `?p=${ page.context.postId }`;
+		if ( page.context?.postType === 'page' ) {
```
Comment (ntsekouras) [inline packages/edit-site/src/store/actions.js:138]: The problem lies to the fact that previous path didn't start with the `/`.

`page.path = `/?p=${ page.context.postId }` should be fine.

Comment (mattwiebe) [reply packages/edit-site/src/store/actions.js:138]: That will not always work, as WP can be installed in the subdirectory of a server, like `example.com/wordpress-installed-here/`.

We would therefore need to qualify the URL against the home URL of the site, which it looks like can be had with `select( 'core' ).getSite().url`.

**Review by mattwiebe — CHANGES_REQUESTED** [review]
I appreciate the quick turnaround on this. With [this comment](https://github.com/WordPress/gutenberg/pull/28413/files#r562773001) addressed, I _think_ this should be good to go.

And while I don't consider this a blocker, I do not really love how a request to `?page_id=n` will result in a `301` redirect on the majority of sites with pretty permalinks enabled. This is part of what I was trying to avoid in #28407, although I failed to mention it (added a comment there)

**Comment by david-szabo97** [comment]
@mattwiebe Thanks for the review! I just pushed a commit which implements an alternative solution. Instead of constructing URL, we fetch the entity record and use the `link` property. Which is the URL to the post / page.

> That will not always work, as WP can be installed in the subdirectory of a server, like example.com/wordpress-installed-here/.
We have another instance where we are using the `path` only.: https://github.com/WordPress/gutenberg/blob/3458607659e7eaefa0de93f7c36702f6b6afce6b/packages/edit-site/src/components/navigate-to-link/index.js#L28

I'll do some testing in a subdirectory and see what happens.

**Review by mattwiebe — APPROVED** [review]
This is a nice change. We could maybe squeeze a bit more efficiency by preloading this entity on the server side in `gutenberg_edit_site_init` [here](https://github.com/WordPress/gutenberg/blob/3e6cf7b0d1c6156d2923bbd009251e678738d5a0/lib/full-site-editing/edit-site-page.php#L118-L127) but I don't see that as a strong need for now.

**Review by vindl — APPROVED** [review]
Worked well for me. :shipit:

**Pull request merged**